docs(tutorial): update to use `v1.5.x` and best practices

gkalpak opened the pull request:
This is a major re-structuring of the tutorial app's codebase, aiming at applying established best practices (in terms of file naming/layout and code organization) and utilizing several new features and enhancements (most notably components) introduced in recent versions of Angular (especially v1.5).

Apart from the overall changes, two new chapters were introduced: one on components and one on code organization.
## 

In the process, several other things were (incidentally) taken care of, including:
- Dependencies were upgraded to latest versions.
- Animations were polished.
- Outdated links were updated.
- The app's base URL was changed to `/` (instead of `/app/`).

BTW, this has been tested with the following versions of Node (on Windows 10) and everything worked fine:
- 0.11.16
- 4.2.6
- 4.4.2
- 5.10.0
## 

This was inspired by (and loosely based on) #13834.
Again, mad props to @teropa for leading the way :)
## 

**Note:**
The old version of the tutorial, that is compatible with Angular version 1.4 or older, has been saved on the `pre-v1.5.0-snapshot` branch of [angular-phonecat](https://github.com/angular/angular-phonecat). The `v1.4.x` version of the tutorial should be pointed to that branch instead of `master`.
## 

Related to angular/angular-phonecat#326.
Related to angular/angular-seed#341.

---

Fixes #12755
Fixes #13312
Fixes #13623
Fixes #13632

Closes #8952
Closes #11726
Closes #12946
Closes #12947
Closes #13198
Closes #13284
Closes #13834
Closes #14178
Closes #14223

gkalpak committed:
docs(tutorial): update to use `v1.5.x` and best practices

This is a major re-structuring of the tutorial app's codebase, aiming at applying established best
practices (in terms of file naming/layout and code organization) and utilizing several new features
and enhancements (most notably components) introduced in recent versions of Angular (especially
v1.5).

Apart from the overall changes, two new chapters were introduced: one on components and one on code
organization.

--
In the process, several other things were (incidentally) taken care of, including:

* Dependencies were upgraded to latest versions.
* Animations were polished.
* Outdated links were updated.
* The app's base URL was changed to `/` (instead of `/app/`).

BTW, this has been tested with the following versions of Node (on Windows 10) and everything worked
fine:

* 0.11.16
* 4.2.6
* 4.4.2
* 5.10.0

--
This was inspired by (and loosely based on) #13834.
Again, mad props to @teropa for leading the way :)

--
**Note:**
The old version of the tutorial, that is compatible with Angular version 1.4 or older, has been
saved on the `pre-v1.5.0-snapshot` branch of
[angular-phonecat](https://github.com/angular/angular-phonecat). The `v1.4.x` version of the
tutorial should be pointed to that branch instead of `master`.

--
Related to angular/angular-phonecat#326.
Related to angular/angular-seed#329.
Related to angular/angular-seed#333.

---
Fixes #12755
Fixes #13312
Fixes #13623
Fixes #13632

Closes #8952
Closes #11726
Closes #12946
Closes #12947
Closes #13198
Closes #13284
Closes #13834
Closes #14178
Closes #14223

gkalpak commented:
This has to get merged simultaneously with corresponding PRs on:

Repo | PRs
--- | ---
_angular-phonecat_ | [#326][phonecat-326]
_angular-seed_ | [#329][seed-329], [#333][seed-333]
_angular.io_ | TBD (see [Upgrading from 1.x][ng2-upgrade])

--
Things that still need to be fixed:

- [ ] The images under `docs/img/tutorial` must be updated to match the new text/code.

--
Things that should/could be cosidered (potentially after this PR has been merged merged):

- [ ] Update the tutorial (and the necessary docs-app directives) on the `v1.4.x` branch, which will still contain the old version of the tutorial, to point to the `pre-v1.5.0-snapshot` branch of [angular-phonecat][phonecat].
- [ ] Adding an "advanced" section at the end, either pointing to resources (e.g. relevant Developer Guide sections) or as subsequent steps. Example topics:
    * Directives/DDO, 2-way/1-way/1-time bindings
    * Forms (+ `ngMessages`)
    * `ngComponentRouter`
    * (multi-slot) transclusion
    * Task runners/Build tools and best practices
    * Testing animations

- [ ] To discuss: Is it better to import the `core` module from the main `phonecatApp` module only, or on a per-view basis ?
- [ ] To discuss: Is it better to split up the CSS per component or keep it in a couple top-level files (e.g. one for animations and one for the rest) ? What about animation-related JS code ?
- [ ] To discuss: Do we want to mention/use Protractor's `directConnect: true` config option ?


[phonecat]: https://github.com/angular/angular-phonecat
[phonecat-326]: https://github.com/angular/angular-phonecat/pull/326
[ng2-upgrade]: https://angular.io/docs/ts/latest/guide/upgrade.html
[seed-329]: https://github.com/angular/angular-seed/pull/329
[seed-333]: https://github.com/angular/angular-seed/pull/333

teropa commented:
@gkalpak I can look into getting angular.io updated. I'll need to look into this branch in more detail to see the extent of the changes need to that guide, but in any case I'll bring it up with the docs team tomorrow.

gkalpak commented:
@teropa: Thx, that would be great ! Also, as you look through it, any feedback is more than welcome :smiley:

BTW, I tried to follow the structure that you used in the Upgrade Guide more closely, so hopefully you will be able to get rid of some of the noise in the guide and focus more on the new/different concepts.

--
Big :+1: for your work on both #13834 and the [Upgrade Guide][1].

[1]: https://angular.io/docs/ts/latest/guide/upgrade.html

petebacondarwin commented on docs/content/tutorial/index.ngdoc:
```diff
@@ -62,24 +65,26 @@ a few git commands.
+access to the `git` command line tool. The main commands that you will need to use are:
+
+* `git clone ...`: Clone a remote repository onto your local machine.
+* `git checkout ...`: Check out a particular branch or a tagged version of the code to hack on.
```
There is an annoying issue about corporate firewalls blocking git

https://github.com/angular/angular-phonecat/issues/141
and probably
https://github.com/angular/angular-phonecat/issues/235

Can we add this information into the guide?

gkalpak replied:
Added

petebacondarwin commented on docs/content/tutorial/index.ngdoc:
```diff
@@ -129,30 +131,32 @@ Once you have Node.js installed on your machine, you can download the tool depen
+* [Protractor][protractor] - end-to-end (E2E) test runner
 
-Running `npm install` will also automatically use bower to download the Angular framework into the
+Running `npm install` will also automatically use bower to download the AngularJS framework into the
```
how about we ditch bower altogether and just use npm to do the installation of angular libraries?

gkalpak replied:
I thought about it, but (afaik) we would need a "build" step that copies the appropriate files from `node_modules/` into `app/`. So, it decided it would be preferrable to keep things simple.

petebacondarwin replied:
We are going to add a `postinstall` hook to copy angular files from `node_modules` to `app/lib`

gkalpak replied:
We decided to go for npm + a `postinstall` script that copies the appropriate files.

gkalpak replied:
In order to avoid the complexity of having a node script that uses some additional module (e.g. `fs-extra`) in order to copy the files etc, we decided to move `index.html` to the root and reference the files inside `node_module/` directly.

petebacondarwin commented on docs/content/tutorial/index.ngdoc:
```diff
@@ -190,103 +194,111 @@ application during development. Start the web server by running:
-If java is already installed and exists in the PATH then you will be shown the version installed,
-if, however you receive a message that "java is not recognized as an internal command or external
-command" you will need to install [java].
+  If JDK is not already install, you can download it [here][jdk-download].
```
If JDK is not already **installed**

gkalpak replied:
Fixed

petebacondarwin commented on docs/content/tutorial/step_00.ngdoc:
```diff
@@ -7,130 +7,143 @@
-<a href="http://localhost:8000/app/" target="_blank" title="Open app on localhost">`http://localhost:8000/app/`</a>
+To see the app running in a browser, open a _separate_ terminal/command line tab or window, then run
+`npm start` to start the web server. Now, open a browser window for the app and navigate to
+http://localhost:8000/index.html.
```
Should we consider using `lite-server` instead of `http-server` since it has a couple of nice features like automatic rewriting to index.html (to support HTML5 mode) and also it starts up the browser when you start the server.

gkalpak replied:
The version of `http-server` we are using now also supports opening the browser.
Since we are not using HTML5 mode in the tutorial (I would rather keep it that way, in order to avoid discussing one more concept (server-side configuration for HTML5 mode)), I don't think `lite-server` offers any advantage over `http-server`. We _could_ switch, but we need to also update `angular-phonecat` and `angular-seed`, so I would avoid it if there is no compelling reason.

How about we do it in a subsequent PR ?

petebacondarwin replied:
OK stick with `http-server`

petebacondarwin replied:
Although lite-server also offers live-reloading :-)

gkalpak replied:
I have lready added this to my post-merge tasks :)

petebacondarwin commented on docs/content/tutorial/step_00.ngdoc:
```diff
@@ -7,130 +7,143 @@
+  <head>
+    <meta charset="utf-8">
+    <title>My HTML File</title>
+    <link rel="stylesheet" href="bower_components/bootstrap/dist/css/bootstrap.css" />
```
`<link>` tags are empty, we should not use a self-closing tag

gkalpak replied:
`<link>` (and `<img>`) are void elements, which according to [the spec][1] are allowed to have a `/` (it's optional for HTML 5). I guess it's a remnant from my XHTML days, but I find it more readable with the slash, so I always use it.

I'm fine removing it, if you feel strongly about it.

[1]: https://www.w3.org/TR/html5/syntax.html#syntax-start-tag

petebacondarwin commented on docs/content/tutorial/step_01.ngdoc:
```diff
@@ -12,11 +12,11 @@ dynamically display the same result with any set of data.
 <div doc-tutorial-reset="1"></div>
 
-
+<br />
```
`<br>` tags are also empty

gkalpak replied:
Same as above.

petebacondarwin commented on docs/content/tutorial/step_02.ngdoc:
```diff
@@ -49,128 +50,150 @@ __`app/index.html`:__
-We replaced the hard-coded phone list with the {@link ng.directive:ngRepeat ngRepeat directive}
-and two {@link guide/expression Angular expressions}:
+We replaced the hard-coded phone list with the {@link ng.directive:ngRepeat ngRepeat} directive and
+two {@link guide/expression Angular expressions}:
```
You should be able to simply write `{@link ngRepeat}` here, unless you don't want the `ngRepeat` to be styled as code, in which case you can write `{@link ngRepeat ntRepeat}`

gkalpak replied:
At first, I styled them as code, but it turned out it was very hard to distinguish clickable code (links) from plain code, so I switched to rely on non-code link styling for clickable/linkable stuff. I will use the simpler form though (`{@link ngRepeat ngRepeat}`).

petebacondarwin commented on docs/content/tutorial/step_03.ngdoc:
```diff
@@ -1,223 +1,279 @@
 
-* Data-binding: This is one of the core features in Angular. When the page loads, Angular binds the
-name of the input box to a variable of the same name in the data model and keeps the two in sync.
+To create a component, we use the `.component()` method of an {@link module Angular module}. We must
```
perhaps a link to the component helper here?

gkalpak replied:
Added

petebacondarwin commented on docs/content/tutorial/step_03.ngdoc:
```diff
@@ -1,223 +1,279 @@
+From the template, we can refer to our controller instance using an alias. This way, the context of
+evaluation for our expressions is even more clear. By default, components use `$ctrl` as the
+controller alias, but we can override it, should the need arise. (It is a very uncommon need though,
+so we will be sticking with `$ctrl` most of the time.)
```
I would not bother with the sentence in parentheses

gkalpak replied:
Removed

petebacondarwin commented on docs/content/tutorial/step_03.ngdoc:
````diff
@@ -1,223 +1,279 @@
+```js
+// Register `phoneList` component, along with its associated controller and template
+angular.
+  module('phonecatApp').
````
is it officially recommended practice to "reopen" the module to add in the component?
Given that you are creating a new file anyway, why not have a one module = one file bijection.
The benefits of this are:
* you don't have to sort the files when building a distribution file
* you can easily move a file to another project without having to bring along its module definition

gkalpak replied:
I don't think it is explicitly mentioned anywhere, but this seems to be what john's styleguide is going throughout.
This is how I do it too fwiw :stuck_out_tongue:

Pros and cons of one module per file:

> you don't have to sort the files when building a distribution file

:+1: 

> you can easily move a file to another project without having to bring along its module definition

The idea is that you are creating a module for the "least viable feature" - i.e. the smaller chunk of functionality that does something worth moving around or re-using in another project. So, there is no point in copying just the controller file of a component over to another project - you would copy the whole component directory instead(including the controller, controller spec, the module definition, the template etc).

> Too many modules

:-1:

> I don't think it's promoted as a good practice anywhere (at least I haven't seen it), so there might be few projects following this convention out there (I'm just speculating here).

:-1: 

--
Could go either way (but only because I'm too tempted by not having to sort the files when bundling/testing :stuck_out_tongue: )
I also have a slight concern about the overhead of creating a module for everything. I think there is a (tiny) bit of work involved in loading a module, which could add up.

Leaving it up to you :grin:

petebacondarwin replied:
Ohhh! It's so hard to decide...

I am generally not in favour of reopening anything. Especially old milk bottles!

So I think I would rather go with one module per file:

* I don't believe that there is not so much of a performance penalty, because there is a similar cost to looking up the module when reopening.
* I believe that having small modules that explicitly declare their dependencies makes them a bit more self documenting.
* My feeling is that if things are so tightly bound that they should be in a single module then they should be in the same file??

If you do feel strongly @gkalpak  then don't do it. But if you are not bothered then I would be delighted to see this changed :-)

petebacondarwin commented on docs/content/tutorial/step_03.ngdoc:
```diff
@@ -1,223 +1,279 @@
 
-* Add an end-to-end test into the `describe` block, `test/e2e/scenarios.js` should look like this:
+    it('should create a `phones` model with 3 phones', inject(function($componentController) {
+      var ctrl = $componentController('phoneList', {$scope: {}});
```
`{$scope: {}}` not needed any more!!

gkalpak replied:
Removed !

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
-two provided sorting options.
+So, now that we learned we should put everything in its own file, our `app/` directory will soon be
+full with dozens of files and specs (remember we keep our unit test files next to the corresponding
+source code files). What's more important, logically related files will not be groupped together; it
```
groupped -> grouped

gkalpak replied:
Fixed

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
-* We then chained the `filter` filter with {@link ng.filter:orderBy `orderBy`}
-filter to further process the input into the repeater. `orderBy` is a filter that takes an input
-array, copies it and reorders the copy which is then returned.
+Well, we are going to group our files into directories _by feature_. For example, since we have a
```
I am pretty sure that directories is a Windows thing :-)
Shouldn't we stick with folders throughout instead?

gkalpak replied:
Hm...based on (unofficial) resources like [this][1], [this][2] or [this][3], I'd say it's definitely not a Windows thing - or at least that `folder` is definitely a Windows thing :stuck_out_tongue: 
In addition, based on the fact that we have CWD (current working directory) and CD (change directory) and not CWF of CF, my understanding is that `directory` is the technically correct term (especially when dealing with command line stuff).

But, I'm actually not too confident on that :stuck_out_tongue:
If you think people will be more familiar with the term `folder`, I'm fine changing it.

[1]: http://superuser.com/questions/169457/directory-vs-folder
[2]: http://superuser.com/questions/187900/what-is-the-difference-between-a-directory-and-a-folder
[3]: http://stackoverflow.com/questions/1191381/english-ui-terminology-directory-or-folder

petebacondarwin replied:
So you went with directory throughout :-) OK

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
-`orderProp` is then used as the input for the `orderBy` filter.
+<div class="alert alert-info">
+  <p>
+    Other typical names for our `core` directory are `common` and `components`. The latter is kind
```
I sometimes use `shared` too

petebacondarwin replied:
what does the "style-guide" have to say about this?

petebacondarwin replied:
at one point people were also playing with the name `ng_modules` or `angular_modules` but this was deemed too verbose.

gkalpak replied:
The de facto [styleguide][1] for ng1 uses `core`, but states that you can choose the name that makes sense to you (e.g. common, shared). Based on recent work on `angular-cli` (which is supposed to be based on the new (unreleased) styleguide), I think `shared` will be the new recommendation.

I chose `core` because that is what was used by @teropa and the guide on [angular.io](https://angular.io).
I don't think it makes much of a difference anyway. Anything is better than `components` at this point :stuck_out_tongue: 

I've added a mention to `shared`.

[1]: https://github.com/johnpapa/angular-styleguide/tree/master/a1#naming

petebacondarwin replied:
`core` is fine

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
````diff
@@ -1,199 +1,310 @@
+      phone-list.component.js
+      phone-list.component.spec.js
+    app.js
+```
````
Perhaps it would be nice for this directory structure to show the `core` folder too?

gkalpak replied:
In this listings (there are more below), I show the current structure after a specific refactoring step (e.g. move feature into own directory, make it its own module, use external template etc).
But even in the final structure shown below, there is no `core/` directory yet - it is createdin the next step, when we create a "shared" feature (the checkmark filter).

I think it would be confusing to show a directory (or folder :stuck_out_tongue:) that we haven't created yet, or to create a directory that we don't need. I can still add it if you feel strongly about it :stuck_out_tongue:

petebacondarwin replied:
OK

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
+## Using Modules
 
+As previously mentioned, one of the benefits of having a modular architecture is code reuse &mdash;
+not only inside the same application, but across application too. There is one final step in making
```
across application**s**

gkalpak replied:
Fixed

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
+    we have created. This might seem tedious, but is totally worth it.
+  </p>
+  <p>
+    In a production-ready application, you are better off concatenating and minifying all your
```
I think this reads slightly better:

> In a production-ready application, you would concatenate and minify all your JavaScript files anyway...

gkalpak replied:
Changed

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
+<div class="alert alert-info">
+  <p>
+    Don't forget to also update your `index.html` adding a `<script>` tag for each JacaScript file
+    we have created. This might seem tedious, but is totally worth it.
```
It might be worth mentioning (since you've done it this way) that the module files must be loaded before component files.

gkalpak replied:
I added a note.

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
+our component definitions cleaner.
 
+So, while it's perfectly fine to keep our component templates inline (using the `template` property
+of the CDO), we decided to use an external template for our `phoneList` component. In order to
```
How about?

> we are going to use

gkalpak replied:
SGTM - changed

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
+component is defined, we place it inside `app/phone-list/`.
 
-Let's turn our attention to the end-to-end test.
+We copied the contents of the `template` property (the HTML code) into
```
Is the language of the tutorial past tense: we did this, we did that?
Or is it imperative: do this or do that?

I am inclined toward the latter.

In which case this should be :

> Copy the contents of the `template` ...

Does that sound right to you? If so can you check this styling throughout?

petebacondarwin replied:
I just realised that perhaps this bit is past tense because each step might already have some files setup for the developer already?

gkalpak replied:
Yes, this is also the current style. At the top of each step, a highlevel overview of the changes are presented along with a link for the full diff on github and then below, we exlain what we did and highlight the most important changed (thus the past tense).

petebacondarwin replied:
OK!

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
+## Final Directory/File Layout
 
-      element(by.model('orderProp')).element(by.css('option[value="name"]')).click();
+After all the refactorings that took place, this is how our application looks like from the outside:
```
I think this should be either

> this is how our application looks from the outside

or

> this is what our application looks like from the outside

I would prefer the former

gkalpak replied:
Changed

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
+      phone-list.template.html
+    app.css
+    app.module.js
+    index.html
```
this is missing the core directory

gkalpak replied:
It's because there isn't one yet (see https://github.com/angular/angular.js/pull/14416/files/0234f66cdb8703393ab24bffc49a467b48e912e7#r61547667).

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
+pass.
+
+<div class="alert alert-success">
+  One of the great things about tests is the confidence the provide, when refactoring your
```
confidence **they** provide

gkalpak replied:
Fixed

petebacondarwin commented on docs/content/tutorial/step_04.ngdoc:
```diff
@@ -1,199 +1,310 @@
+<div class="alert alert-success">
+  One of the great things about tests is the confidence the provide, when refactoring your
+  application. It's easy to break something as you start moving files around and re-arranging
+  modules. Having a good test coverage is the quickest, easiest and most reliable way of knowing
```
> Having good test coverage

gkalpak replied:
Fixed

petebacondarwin commented on docs/content/tutorial/step_05.ngdoc:
```diff
@@ -1,277 +1,168 @@
 
-The `$` prefix is there to namespace Angular-provided services.
-To prevent collisions it's best to avoid naming your services and models anything that begins with a `$`.
+* Use of the `filter` filter: The {@link ng.filter:filter filter} function uses the `$ctrl.query`
```
pretty sure this link can be more simply: `{@link filter}`

gkalpak replied:
In this case, we need the `ng.filter:` part, because there is also `guide/filter`, so dgeni doesn't know what to pick for `filter`. (And as mentioned above, I chose not to format links as code for "distinguishability".)

petebacondarwin replied:
OK

petebacondarwin commented on docs/content/tutorial/step_05.ngdoc:
```diff
@@ -1,277 +1,168 @@
-dependency injector would not be able to identify services correctly.
+In previous steps, we learned how to write and run unit tests. Unit tests are perfect for testing
+controllers and other parts of our application written in JavaScript, but they can't easily
+test DOM manipulation or the wiring of our application. For these, an end-to-end (E2E) test is a
```
might be worth mentioning templates here:

> can't easily test templates, DOM manipulation or wiring of our application

petebacondarwin replied:
actually unless you defined wiring earlier on, I think this is a bit of a vague term. Perhaps `bindings` would be better?

gkalpak replied:
Good point. (Since `bindings` sounded more narrow-scoped than intended,) I changed it to: 

> can't easily test templates, DOM manipulation or interoperability of components and services

WDYT ?

petebacondarwin commented on docs/content/tutorial/step_05.ngdoc:
```diff
@@ -1,277 +1,168 @@
-
-<pre>Chrome 22.0: Executed 2 of 2 SUCCESS (0.028 secs / 0.007 secs)</pre>
+Much like Karma is the test runner for unit tests, we use Protractor to run E2E tests. Try it with
+`npm run protractor`. E2E tests are slow, so unlike with unit tests, Protractor will exit after the
```
> E2E tests take time

Sounds less negative :-)

gkalpak replied:
The glass is half-full, right :smiley:
Changed

petebacondarwin commented on docs/content/tutorial/step_05.ngdoc:
```diff
@@ -1,277 +1,168 @@
 
+<div class="alert alert-info">
+  **Note:** You must ensure your application is being served via a web server to test with
+  protractor. You can do this by using `npm start`.
```
In a different terminal/console

gkalpak replied:
I changed the whole message to:

>   **Note:** In order for protractor to access and run tests against your application, it must be
  served via a web server. In a different terminal/command line window, run `npm start` to fire up
  the web server.

WDYT ?

petebacondarwin commented on docs/content/tutorial/step_06.ngdoc:
```diff
@@ -1,108 +1,239 @@
-an invalid URL `http://localhost:8000/app/{{phone.imageUrl}}`, which it would have done if we had
-only specified an attribute binding in a regular `src` attribute (`<img src="{{phone.imageUrl}}">`).
-Using the `ngSrc` directive prevents the browser from making an http request to an invalid location.
+* We then chained the `filter` filter with the {@link ng.filter:orderBy orderBy} filter to further
```
`{@link orderBy}`?

gkalpak replied:
Removed the `ng.filter:` prefix.

petebacondarwin commented on docs/content/tutorial/step_06.ngdoc:
```diff
@@ -1,108 +1,239 @@
+    var ctrl;
+
+    beforeEach(inject(function($componentController) {
+      ctrl = $componentController('phoneList', {$scope: {}});
```
No need for the `{$scope: {}}` any more :-)

gkalpak replied:
Removed !

petebacondarwin commented on docs/content/tutorial/step_06.ngdoc:
```diff
@@ -1,108 +1,239 @@
+          'MOTOROLA XOOM\u2122'
+        ]);
+
+        element(by.model('$ctrl.orderProp')).element(by.css('option[value="name"]')).click();
```
Is it worth extracting out the protractor queries here:

```
var orderSelect = element(by.model('$ctrl.orderProp'));
var nameOption = orderSelect.element(by.css('option[value="name"]'));
nameOption.click();
```

gkalpak replied:
Done

petebacondarwin commented on docs/content/tutorial/step_06.ngdoc:
```diff
@@ -1,108 +1,239 @@
-or Chrome's Web Inspector, or inspecting the webserver access logs, confirm that the app is indeed
-making an extraneous request to `/app/%7B%7Bphone.imageUrl%7D%7D` (or
-`/app/{{phone.imageUrl}}`).
+<div></div>
```
Is this to force some whitespace?

gkalpak replied:
`<ul>` right after a heading (`# Experiments`) have some different styling. Inserting a `<div>` to avoid `h1 + ul` selectors (iirc). Not ideal, I know :smiley:

petebacondarwin replied:
OK

petebacondarwin commented on docs/content/tutorial/step_07.ngdoc:
```diff
@@ -1,380 +1,311 @@
-multiple views by adding routing, using an Angular module called 'ngRoute'.
+Enough of building an app with three phones in a hard-coded dataset! Let's fetch a larger dataset
+from our server using one of Angular's built-in {@link guide/services services} called
+{@link ng.$http $http}. We will use Angular's {@link guide/di dependency injection (DI)} to provide
```
`{@link ng.$http}` will format the `$http` as an inline code block. As you have it now the $http will look like normal text. What were you aiming for?

gkalpak replied:
Normal (link) text :smiley:

petebacondarwin commented on docs/content/tutorial/step_07.ngdoc:
```diff
@@ -1,380 +1,311 @@
-unless it is configured with proper module definitions.
+The `$http` service returns a {@link ng.$q promise object}, which has a `then()` method. We call
+this method to handle the asynchronous response and assign the phone data to the controller, as a
+model called `phones`. Notice that Angular detected the JSON response and parsed it for us into the
```
model -> property ?

gkalpak replied:
This is what they are called in the current version of the tutorial (I guess trying to make the connection with the MVC paradigm more obvious).
Changed it to `property`.

petebacondarwin commented on docs/content/tutorial/step_07.ngdoc:
```diff
@@ -1,380 +1,311 @@
+`data` property of the `response` object passed to our callback!
 
-The injector only carries out the following steps :
+Since we are making the assignment of the `phones` model in a callback function, where the `this`
```
model -> property

gkalpak replied:
Changed.

petebacondarwin commented on docs/content/tutorial/step_07.ngdoc:
```diff
@@ -1,380 +1,311 @@
 
-Notice the second argument passed to `angular.module`, `['ngRoute', 'phonecatControllers']`. This
-array lists the modules that `phonecatApp` depends on.
+From this point onwards, we are going to use the inline method in the tutorial. With that in mind,
```
When defining (and annotating) controllers inline, does that make putting methods on the controller's prototype more tricky?

petebacondarwin replied:
Do we want people to start thinking about component controllers more as classes than constructor functions?

gkalpak replied:
I don't think annotation style makes much difference. Inline definition does indeed make it more difficult to put stuff on the prototype, but refactoring them is trivial as well: E.g.:

```js
.component('myComp', {
  templateUrl: '...',
  bindings: {...},
  controller: function MyCompController(...) {
    this.someProp = ...;
    this.someMethod = ...;
    ...
  }
})

// to

.component('myComp', {
  templateUrl: '...',
  bindings: {...},
  controller: MyCompController
})

function MyCompController(...) {
  this.someProp = ...;
  ...
}
MyCompController.prototype.someMethod = ...;
...
```

--
Personally, I don't see much benefit in using the prototype for controllers and services. And by the time people get to bother with the difference of defining methods in the constructor vs on the prototype, classes will be all over the place :smiley:

I like how the inline definition looks much more similar to a class (with TypeScript annotation in ng2). I.e.:

```js
.component('myComp', {
  templateUrl: '...',
  bindings: {...},
  controller: function MyCompController(...) {
    this.someProp = ...;
    this.someMethod = ...;
    ...
  }
})

// looks at lot like:

@Component(
  selector: 'myComp',
  templateUrl: '...',
  inputs: [...]
}
class MyCompController {
  this.someProp = ...;
  someMethod() {...}
  ...
}
```

--
Most importantly, for a tutorial, I believe simpler is better.
(Plus, I could definitely do without having to exlain JS prototypal inheritance :grin:)

petebacondarwin replied:
OK. Sounds good!

petebacondarwin commented on docs/content/tutorial/step_07.ngdoc:
```diff
@@ -1,380 +1,311 @@
+  {@link ngMock.$httpBackend mock version} of the service that in a production environment
+  facilitates all XHR and JSONP requests. The mock version of this service allows us to write tests
+  without having to deal with native APIs and the global state associated with them — both of which
+  make testing a nightmare.
```
And also overcomes the asynchronous nature of these calls, which slows down unit tests

gkalpak replied:
Added

petebacondarwin commented on docs/content/tutorial/step_08.ngdoc:
```diff
@@ -1,197 +1,119 @@
+...
+<ul class="phones">
+  <li ng-repeat="phone in $ctrl.phones | filter:$ctrl.query | orderBy:$ctrl.orderProp" class="thumbnail">
+    <a href="#/phones/{{phone.id}}" class="thumb">
```
the default hash prefix is now `!` so these links should probably be: `#!/phones/{{phone.id}}`

gkalpak replied:
It's not the default on `1.5.x` that the tutorial is using. Something for the next major tutorial update once `1.6.0` is out :stuck_out_tongue:

petebacondarwin replied:
Oh! Right.
Perhaps we should set the value to `"!"` in a config block in this tutorial?

petebacondarwin commented on docs/content/tutorial/step_08.ngdoc:
```diff
@@ -1,197 +1,119 @@
-      expect(element(by.binding('phone.name')).getText()).toBe('Nexus S');
+      element.all(by.css('.phones li a')).first().click();
+      browser.getLocationAbsUrl().then(function(url) {
+        expect(url).toBe('/phones/nexus-s');
```
I think that `getLocationAbsUrl` will actually return something with `http://localhost...` etc at the beginning

gkalpak replied:
Based on the fact that the test passes, it doesn't :stuck_out_tongue: 
I think there might be a bug in protractor: https://github.com/angular/protractor/blob/7b0302c952f200231957c3857050e0fa8d5b09d3/lib/clientsidescripts.js#L651-L658

If there is `angular.getTestability()`, is returns `$$tstability`'s `getLocation()` (which returns `$location.url()`).
If there isn't `angular.getTestability()`, then it returns `$locaton.absUrl()`. I don't know under what circumstances there might be no `angular.getTestability()` though.

gkalpak replied:
It seems that it's been like this since Protractor `v1.1.1` (unless we change what `$$testability.getLocation()` returns in core): https://github.com/angular/protractor/commit/0e9c2f49dbe8cdbe56a2ce270f504e26cea0a111

gkalpak replied:
No, it's been like this since the beginning of `$$testability`: https://github.com/angular/angular.js/commit/85880a64900fa22a61feb926bf52de0965332ca5

gkalpak replied:
I submitted angular/protractor#3185

petebacondarwin replied:
OK. So if that issue gets resolved will we have to change our tests here?

gkalpak replied:
If it gets resolved it will be a BC for Protractor (so it has to come with a major version update).
If we update to that major version, then yes, we will need to update our tests.

petebacondarwin commented on docs/content/tutorial/step_09.ngdoc:
```diff
@@ -1,144 +1,414 @@
-certain phone features were present or not. We have used a custom filter to convert those text
-strings into glyphs: ✓ for "true", and ✘ for "false". Let's see what the filter code looks like.
+In this step, you will learn how to create a layout template and how to build an application that
+has multiple views by adding routing, using an Angular module called {@link ngRoute ngRoute}.
```
this form of the link will not format the ngRoute as code block. Is that what you want?

gkalpak replied:
Yup :smiley:

petebacondarwin commented on docs/content/tutorial/step_09.ngdoc:
```diff
@@ -1,144 +1,414 @@
+
+* Load the module definition(s) that you specify in your application.
+* Register all Providers defined in these module definition(s).
+* When asked to do so, inject a specified function and any necessary dependencies (services), that
```
This doesn't read that well. Perhaps...

> When asked to do so, lazily instantiate services and their dependencies, via their Providers, a parameters to an injectable function.

gkalpak replied:
Changed

petebacondarwin commented on docs/content/tutorial/step_09.ngdoc:
```diff
@@ -1,144 +1,414 @@
+  `$routeProvider` into `PhoneListController` at runtime.
+</div>
+
+Angular modules solve the problem of removing global state from the application and provide a way
```
global variables ?

gkalpak replied:
"State" sounds cooler, but changed it anyway :stuck_out_tongue:

petebacondarwin replied:
There is plenty of global state in an Angular app even without global variables :-)
So you've still got plenty of cool

petebacondarwin commented on docs/content/tutorial/step_09.ngdoc:
```diff
@@ -1,144 +1,414 @@
-describe('filter', function() {
+    files: [
+      'bower_components/angular/angular.js',
+      'bower_components/angular-resource/angular-resource.js',
```
should be

```
'bower_components/angular-resource/angular-route.js',
```

right?

gkalpak replied:
Right :smiley:

petebacondarwin commented on docs/content/tutorial/step_09.ngdoc:
```diff
@@ -1,144 +1,414 @@
+  it('should redirect `index.html` to `index.html#/phones', function() {
+    browser.get('index.html');
+    browser.getLocationAbsUrl().then(function(url) {
+      expect(url).toBe('/phones');
```
absolute URL?

gkalpak replied:
See https://github.com/angular/angular.js/pull/14416/files/0234f66cdb8703393ab24bffc49a467b48e912e7#r61557295.

petebacondarwin commented on docs/content/tutorial/the_end.ngdoc:
```diff
@@ -9,11 +9,18 @@ previous steps using the `git checkout` command.
-We hope this tutorial was useful to you and that you learned enough about Angular to make you want
-to learn more. We especially hope you are inspired to go out and develop Angular web apps of your
-own, and that you might be interested in {@link misc/contribute contributing} to Angular.
+We hope this tutorial was useful to you and that you learned enough about AngularJS to make you want
```
Angular

gkalpak replied:
I used `AngularJS` when I felt that the framework as a whole was referred to or where a distinction with Angular2 (or AngularDart) should be made and used just `Angular` in all other places. (FWIW, I have thought about every occurence and made a conscious choice for `Angular` vs `AngularJS` - "conscious" doesn't necessarily mean "most appropriate" :stuck_out_tongue: )

If you think it's better to use `Angular` thoughout, I'm fine as well.

petebacondarwin replied:
It's just that this is what I was told to do by a Googler... possibly Igor or Naomi.

petebacondarwin replied:
Leave them as you have them and see if anyone on the Google team notices :-) (and complains)

petebacondarwin commented on docs/content/tutorial/the_end.ngdoc:
```diff
@@ -9,11 +9,18 @@ previous steps using the `git checkout` command.
-own, and that you might be interested in {@link misc/contribute contributing} to Angular.
+We hope this tutorial was useful to you and that you learned enough about AngularJS to make you want
+to learn more. We especially hope you are inspired to go out and develop Angular web applications of
+your own, and that you might be interested in {@link misc/contribute contributing} to AngularJS.
```
I think we should use Angular throughout

gkalpak replied:
See above.

petebacondarwin replied:
Seen

petebacondarwin commented on docs/content/tutorial/step_14.ngdoc:
```diff
@@ -0,0 +1,565 @@
+[caniuse-css-transitions]: http://caniuse.com/#feat=css-transitions
+[jquery]: https://jquery.com/
+[jquery-animate]: https://api.jquery.com/animate/
+[webplatform-animations]: https://docs.webplatform.org/wiki/css/properties/animations
```
I assume that you didn't make any significant changes to this module?

gkalpak replied:
Define "significant" :stuck_out_tongue: 
IIRC I changed the wording here and there, added some experiments and removed CSS prefixes that aren't necessary when using latest versions of browsers (I left a note that older browsers might need prefixes).

teropa commented on docs/content/tutorial/step_13.ngdoc:
````diff
@@ -0,0 +1,321 @@
+```
+
+<br />
+**`app/core/phone/phone.service.js`:**
````
Should it be `phone.factory.js`?

petebacondarwin replied:
IMO it should be `service` since that it what it is, in the end. The fact that we just the `factory()` helper to define the service is an implementation detail that it not what we are trying to convey here.

teropa replied:
OK, makes sense. johnpapa/angular-styleguide uses `.factory.js` a lot but I think I actually like this more.

gkalpak committed:
docs(tutorial): update images to match new content

gkalpak committed:
fixup 1

- Fix typos.
- Improve wording.
- Simplify links.
- Add warnings:
  - Firewalls blocking `git:`.
  - File include order (i.e. `.module` files before other files).
- Improve tests:
  - Remove unnecessary locals when calling `$componentController()`.
  - Extract protractor queries.

gkalpak commented:
I added three more commits:

- 9e13bd0: Merged a commit from @petebacondarwin that added SVG format for images.
- d68265f: Updated the tutorial images to match the new content of the tutorial (also fixed up and re-instated 3 hidden images).
- 0002c05: Addressed feedback/comments on the original commit. The equivalent changes have been made to  angular/angular-phonecat#326 (where necessary).

@petebacondarwin, PTAL :smiley:

petebacondarwin commented:
@gkalpak OK, so I have run through all our comment conversations.
I will do a walk through of the actual tutorial tomorrow and see if it all makes sense.
Thanks for all your work on this!

petebacondarwin commented:
I found the bit about `git://` and proxies. Great! I think it would benefit from a bit of a heading or title in the alert box, to make it easier to find if you are having problems. Perhaps : **Firewall/Proxy Issues**?

petebacondarwin commented:
It may be worth putting another alert box after the first `npm start` explaining what the error message means if you already have a server running on port 8000

petebacondarwin commented:
It might be worth saying that updating webdriver can take quite a long time (minutes on my machine) and if you Ctrl-C out of it then you get weird errors when trying to run protractor later.

petebacondarwin commented:
In step-5 the protractor tests fail because they are trying to access `localhost:8000/index.html` whereas the correct url is `localhost:8000/app/index.html`.

I believe the simplest fix is to change the `browser.get` to:

```
    beforeEach(function() {
      browser.get('app/index.html');
    });
```

Another alternative fix is to change the baseUrl in the protractor config:

```
  baseUrl: 'http://localhost:8000/app/',
```

petebacondarwin commented:
I take that last comment back. I see that `npm start` has changed to run the local web server from the `app` folder. So this is fine

petebacondarwin commented:
In step-6 we need to change the headings from "template" and "controller" to "component template" and "component controller" in keeping with step-5

petebacondarwin commented:
Similarly in step-7 the "controller" heading should change to "component", I believe

petebacondarwin commented:
And for subsequent steps

petebacondarwin commented:
By the way I love the new images but perhaps the gradient fills are a bit cheesy?

petebacondarwin commented:
The next step button on step-12 takes you to the end.

gkalpak committed:
fixup 2

- Fix typos.
- Fix navigation for steps > 12.
- Make section title consistent (include "Component" where appropriate).
- Improve alert formatting (titles, line spacing).

gkalpak commented:
I added a new commit that should address all above comments except for the image gradient cheesy-ness (because I am not sure what it means). Is it [like this][1] or like too yellow ?

[1]: http://el.urbandictionary.com/define.php?term=cheesy

petebacondarwin commented:
By cheesy I meant a bit 1980s in style. I realise that you were following the form of the current app.
But I think it would benefit with a flatter look.

petebacondarwin commented on docs/content/tutorial/step_05.ngdoc:
```diff
@@ -141,6 +141,12 @@ To rerun the test suite, execute `npm run protractor` again.
+  another instance of a running web server) you will get an `EADDRINUSE` error. Make sure the port 
+  is available before running `npm start`.
+</div>
+
```
I think this warning should also appear in the introduction when setting up the environment. At least that is where I encountered this error.

gkalpak replied:
I put it here, because we already have 8-9 alerts in the introduction. I can move it there if think it's better. (One more won't hurt I guess :smiley:)

gkalpak replied:
Do you think would be better to move all the warnings in the introduction in a separate sections (at the bottom of introduction).

petebacondarwin replied:
That might work. A kind of setup troubleshooting section?

gkalpak committed:
fixup 3

- Fix typos.
- Improve common issues descriptions.
- Move common setup issues to dedicated section.

gkalpak committed:
fixup 4

- Use `!` as hash-prefix.

gkalpak commented:
Added a couple more fixups that move warning in a "Common Issues" section (in the introduction) and use `!` as hash-prefix (angular-phonecat and angular-seed have been updated accordingly).
/cc @petebacondarwin

gkalpak commented:
Tasks for after getting this is:

- [ ] Make appropriate modifications to the 1.4.x tutorial (see https://github.com/angular/angular.js/pull/14416#issuecomment-208844996).
- [ ] Use a more flat look in images - remove gradient.
- [ ] Switch to `lite-server` (with auto-reloading).

petebacondarwin commented:
LGTM
